File: 2015/python/d06.py
TOGGLE = '0'
TURN_OFF = '1'
TURN_ON = '2'

def read(filename):
    with open(filename, 'r') as file:
        return [line.replace('\n', '').
                     replace('toggle ', TOGGLE + ',').
                     replace('turn off ', TURN_OFF + ',').
                     replace('turn on ', TURN_ON + ',').
                     replace(' through ', ',').split(',') for line in file]

def update1(t, x, y, grid):
  if (t == TOGGLE):
    grid[y][x] = 1 - grid[y][x]
  elif (t == TURN_ON):
    grid[y][x] = 1
  else:
    grid[y][x] = 0

def update2(t, x, y, grid):
  if (t == TOGGLE):
    grid[y][x] += 2
  elif (t == TURN_ON):
    grid[y][x] += 1
  else:
    grid[y][x] = max(0, grid[y][x] - 1)

def updateall(lines, grid, part2 = False):
    for d in lines:
        (t, x1, y1, x2, y2) = d[0], int(d[1]), int(d[2]), int(d[3]), int(d[4])
        for x in range(x1, x2 + 1):
            for y in range(y1, y2 + 1):
                if part2:
                    update2(t, x, y, grid)
                else:
                    update1(t, x, y, grid)

    return sum(map(sum, grid))

def part1(lines):
    grid = [[0 for x in range(1000)] for y in range(1000)]
    return updateall(lines, grid)

def part2(lines):
    grid = [[0 for x in range(1000)] for y in range(1000)]
    return updateall(lines, grid, True)

d = read('../inputs/input_6.txt')
print(part1(d))
print(part2(d))

File: 2015/R/d06.R
d <- read.csv(text = unlist(lapply(readLines("../inputs/input_6.txt"), 
  function(x) gsub(" through ", ",", gsub("turn off ", "off,", 
                gsub("turn on ", "on,", gsub("toggle ", "tog,", x)))))), 
                  header = FALSE, col.names = c("f","x1","y1","x2","y2"))
d[, c("x1", "x2", "y1", "y2")] <- d[, c("x1", "x2", "y1", "y2")] + 1

light <- function(d, tog, on, off) {
  m <- matrix(nrow = 1000, ncol = 1000, data = 0)
  for (i in seq_len(nrow(d))) {
    m[d$y1[i]:d$y2[i], d$x1[i]:d$x2[i]] <- 
      do.call(d$f[i], list(m[d$y1[i]:d$y2[i], d$x1[i]:d$x2[i]]))
  }
  sum(m)
}

c(light(d, function(x) (1 - x), function(x) 1, function(x) 0),
  light(d, function(x) (2 + x), function(x) (x + 1), function(x) pmax(0, x - 1)))
